docker/archive/dest.go

```diff
@@ -0,0 +1,60 @@
+package archive
+
+import (
+	"io"
+	"os"
+
+	"github.com/containers/image/docker/tarfile"
+	"github.com/containers/image/types"
+	"github.com/pkg/errors"
+)
+
+type archiveImageDestination struct {
+	*tarfile.Destination // Implements most of types.ImageDestination
+	ref                  archiveReference
+	writer               io.Closer
+}
+
+func newImageDestination(ctx *types.SystemContext, ref archiveReference) (types.ImageDestination, error) {
+	if ref.destinationRef == nil {
+		return nil, errors.Errorf("docker-archive: destination reference not supplied (must be of form <path>:<reference:tag>)")
+	}
+	fh, err := os.OpenFile(ref.path, os.O_WRONLY|os.O_EXCL|os.O_CREATE, 0644)
+	if err != nil {
+		// FIXME: It should be possible to modify archives, but the only really
+		//        sane way of doing it is to create a copy of the image, modify
+		//        it and then do a rename(2).
+		if os.IsExist(err) {
+			err = errors.New("docker-archive doesn't support modifying existing images")
+		}
+		return nil, err
+	}
+
+	return &archiveImageDestination{
+		Destination: tarfile.NewDestination(fh, ref.destinationRef),
+		ref:         ref,
+		writer:      fh,
+	}, nil
+}
+
+// Reference returns the reference used to set up this destination.  Note that this should directly correspond to user's intent,
+// e.g. it should use the public hostname instead of the result of resolving CNAMEs or following redirects.
+func (d *archiveImageDestination) Reference() types.ImageReference {
+	return d.ref
+}
+
+// Close removes resources associated with an initialized ImageDestination, if any.
+func (d *archiveImageDestination) Close() {
+	d.writer.Close()
+}
+
+// Commit marks the process of storing the image as successful and asks for the image to be persisted.
+// WARNING: This does not have any transactional semantics:
+// - Uploaded data MAY be visible to others before Commit() is called
+// - Uploaded data MAY be removed or MAY remain around if Close() is called without Commit() (i.e. rollback is allowed but not guaranteed)
+func (d *archiveImageDestination) Commit() error {
+	if err := d.Destination.Commit(); err != nil {
+		return err
+	}
+	return d.writer.Close()
```

Comment:
(The `.Commit()` and `.Close()` _aren’t_ meant to be mutually exclusive. `Close()` destroys temporary state like file handles and temporary files; `Commit()` tells the destination “now you have everything, you can mark the image as finished / start transmitting the result as a single unit”.

See how `daemonImageDestination` terminates its helper goroutine in `Close()` (to ensure that it never lingers around), but correctly closes the generated tar file (making it acceptable to `dockerd` so that the input isn’t all thrown away) in `Commit()`.

The expected pattern for a writer is to do

```go
dest, … := someRef.NewImageDestination()
defer dest.Close() // Or perhaps a more thorough variant with error propagation
// steps which may fail, e.g. reading input
… dest.PutBlob(…)
// other steps which may fail, including other writes to dest
// we get here only if everything has succeeded so far
err = dest.Commit()
```
i.e. the caller can do a blind `defer dest.Close()`. If the two were expected to be exclusive, the calller would have to do something like
```go
succeeded = false
dest, … : someRef.NewImageDestination()
defer func() {
  if succeeded { dest.Commit() } else { dest.Close() }
}
// steps which may fail, e.g. reading input
… dest.PutBlob(…)
// other steps which may fail, including other writes to dest
// we get here only if everything has succeeded so far
succeeded = true
```

which could make the `ImageDestination` implementations simpler, but it would be notably more cumbersome for the caller. Especially because so many `ImageDestination` have no state to speak of and their `.Close()` is empty, it seems overall simpler to force the `ImageDestination` implementations—only those who care—to track success/failure, like `daemonImageDestination.committed` does, than to force all writers to track this.

Perhaps I should make the comments in `types.go` documenting `ImageDestination` and `ImageDestination.Close` more explicit about this?)